### DISCUSSION TITLE
Migration from react-router

### DISCUSSION THREAD

**Discussion by hverlin** [TanStack/router#891] [Q&A]
What's the best strategy to incrementally migrate from react-router?

**Comment by nucleartux** [comment]
I have the same question. I have more than 1000 routes, and I'm not sure if it's possible to migrate gradually.

**Comment by hverlin** [comment]
There is a guide in the docs now: https://tanstack.com/router/v1/docs/framework/react/migrate-from-react-router (added in https://github.com/TanStack/router/pull/1072 by @Benanna2019) 

> At this point I don’t know if you can do a gradual migration, but it seems likely you could have multiple router providers, not desirable.
> The api’s between React Router and Tanstack Router are very similar and could most likely be handled in a sprint cycle or two if that is your companies way of doing things.

**Reply by Benanna2019** [reply]
@nucleartux the primary issue I believe you will run into in a gradual migration, or the primary issue I have run into in building an example, is having to do some recon to figure out if I have routes that are stand alone and then grouping routes together with how much navigation I am doing in the app across routes. 

So for instance, the demo app I have has a products, products/$productId, sell, orders, orders/$orderId and a few other routes. 

What you will run into, like I am, is if I just want to move a subset of routes to a new router, but those routes have navigation within them to routes in another router, I will get errors due to route context and api's being used outside a router context. 

Unless you go the route of building two separate route trees, I think, for an app with more than 1000 routes, find what can be chunked into a new router, and gradually move that over. 

If I come up with something better or if I find out I am making this more difficult than need be, I will follow back up here.

**Reply by hverlin** [reply]
This is what I have seen as well. You need to have both react router and tan-stack router present, which does not seem possible at the moment. In my case, I need to migrate everything at once, which is quite complicated (in particular, converting all the search params handling).
